pkg/indexers/servicemonitor_index.go

```diff
@@ -0,0 +1,200 @@
+		raw, err := ri.index.GetInternal(key)
+		if err != nil {
+			return err
+		}
+		if len(raw) > 0 {
+			var data prom.ServiceMonitorList
+			err := json.Unmarshal(raw, &data)
+			if err != nil {
+				return err
+			}
+			var monitors []*prom.ServiceMonitor
+			for i, valueSvc := range data.Items {
+				if ri.equal(svcMonitor, valueSvc) {
+					monitors = append(data.Items[:i], data.Items[i+1:]...)
+					break
+				}
+			}
+
+			if len(monitors) == 0 {
+				// Remove unnecessary index
+				err = ri.index.DeleteInternal(key)
+				if err != nil {
+					return err
+				}
+			} else {
+				raw, err := json.Marshal(prom.ServiceMonitorList{Items: monitors})
+				if err != nil {
+					return err
+				}
+				err = ri.index.SetInternal(key, raw)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (ri *ServiceMonitorIndexerImpl) Update(old, new *prom.ServiceMonitor) error {
+	if !reflect.DeepEqual(old.Spec.Selector, new.Spec.Selector) {
+		// Only update if selector changes
+		err := ri.Delete(old)
+		if err != nil {
+			return err
+		}
+		err = ri.Add(new)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ri *ServiceMonitorIndexerImpl) serviceForServiceMonitors(svcMonitor *prom.ServiceMonitor) (*apiv1.ServiceList, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&svcMonitor.Spec.Selector)
+	if err != nil {
+		return &apiv1.ServiceList{}, err
+	}
+	if svcMonitor.Spec.NamespaceSelector.Any {
+		return ri.kubeClient.CoreV1().Services(metav1.NamespaceAll).List(metav1.ListOptions{
+			LabelSelector: selector.String(),
+		})
+	}
+
+	list := &apiv1.ServiceList{Items: make([]apiv1.Service, 0)}
+	for _, ns := range svcMonitor.Spec.NamespaceSelector.MatchNames {
+		pods, err := ri.kubeClient.CoreV1().Services(ns).List(metav1.ListOptions{
+			LabelSelector: selector.String(),
+		})
+		if err == nil {
+			list.Items = append(list.Items, pods.Items...)
+		}
+	}
+	return list, nil
+}
+
+func (ri *ServiceMonitorIndexerImpl) equal(a, b *prom.ServiceMonitor) bool {
+	if a.Name == b.Name && a.Namespace == b.Namespace {
+		return true
+	}
+	return false
+}
+
+func (ri *ServiceMonitorIndexerImpl) Key(meta metav1.ObjectMeta) []byte {
+	return []byte(util.GetGroupVersionKind(&apiv1.Service{}).String() + "/" + meta.Namespace + "/" + meta.Name)
+}
+
+func (ri *ServiceMonitorIndexerImpl) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	log.Infoln("Request received at", req.URL.Path)
+	params, found := pat.FromContext(req.Context())
+	if !found {
+		http.Error(w, "Missing parameters", http.StatusBadRequest)
+		return
+	}
+
+	namespace, name := params.Get(":namespace"), params.Get(":name")
+	if len(namespace) > 0 && len(name) > 0 {
+		key := ri.Key(v1.ObjectMeta{Name: name, Namespace: namespace})
+		if val, err := ri.index.GetInternal(key); err == nil && len(val) > 0 {
+			if err := json.NewEncoder(w).Encode(json.RawMessage(val)); err == nil {
+				w.Header().Set("Content-Type", "application/json")
```

Comment:
Also, "x-content-type-options": "nosniff", // http://stackoverflow.com/a/3146618/244009

In the other HTTP handlers, too